In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('../data/raw/DataCoSupplyChainDataset.csv', encoding = 'latin-1')

In [3]:
df.head()

,Type,Days for shipping (real),Days for shipment (scheduled),Benefit per order,Sales per customer,Delivery Status,Late_delivery_risk,Category Id,Category Name,Customer City,...,Order Zipcode,Product Card Id,Product Category Id,Product Description,Product Image,Product Name,Product Price,Product Status,shipping date (DateOrders),Shipping Mode
0,DEBIT,3,4,91.250000,314.640015,Advance shipping,0,73,Sporting Goods,Caguas,...,NaN,1360,73,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,2/3/2018 22:56,Standard Class
1,TRANSFER,5,4,-249.089996,311.359985,Late delivery,1,73,Sporting Goods,Caguas,...,NaN,1360,73,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,1/18/2018 12:27,Standard Class
2,CASH,4,4,-247.779999,309.720001,Shipping on time,0,73,Sporting Goods,San Jose,...,NaN,1360,73,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,1/17/2018 12:06,Standard Class
3,DEBIT,3,4,22.860001,304.809998,Advance shipping,0,73,Sporting Goods,Los Angeles,...,NaN,1360,73,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,1/16/2018 11:45,Standard Class
4,PAYMENT,2,4,134.210007,298.250000,Advance shipping,0,73,Sporting Goods,Caguas,...,NaN,1360,73,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,1/15/2018 11:24,Standard Class


In [4]:
pd.set_option('display.max_columns', None)

In [5]:
df.head()

,Type,Days for shipping (real),Days for shipment (scheduled),Benefit per order,Sales per customer,Delivery Status,Late_delivery_risk,Category Id,Category Name,Customer City,Customer Country,Customer Email,Customer Fname,Customer Id,Customer Lname,Customer Password,Customer Segment,Customer State,Customer Street,Customer Zipcode,Department Id,Department Name,Latitude,Longitude,Market,Order City,Order Country,Order Customer Id,order date (DateOrders),Order Id,Order Item Cardprod Id,Order Item Discount,Order Item Discount Rate,Order Item Id,Order Item Product Price,Order Item Profit Ratio,Order Item Quantity,Sales,Order Item Total,Order Profit Per Order,Order Region,Order State,Order Status,Order Zipcode,Product Card Id,Product Category Id,Product Description,Product Image,Product Name,Product Price,Product Status,shipping date (DateOrders),Shipping Mode
0,DEBIT,3,4,91.250000,314.640015,Advance shipping,0,73,Sporting Goods,Caguas,Puerto Rico,XXXXXXXXX,Cally,20755,Holloway,XXXXXXXXX,Consumer,PR,5365 Noble Nectar Island,725.0,2,Fitness,18.251453,-66.037056,Pacific Asia,Bekasi,Indonesia,20755,1/31/2018 22:56,77202,1360,13.110000,0.04,180517,327.75,0.29,1,327.75,314.640015,91.250000,Southeast Asia,Java Occidental,COMPLETE,NaN,1360,73,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,2/3/2018 22:56,Standard Class
1,TRANSFER,5,4,-249.089996,311.359985,Late delivery,1,73,Sporting Goods,Caguas,Puerto Rico,XXXXXXXXX,Irene,19492,Luna,XXXXXXXXX,Consumer,PR,2679 Rustic Loop,725.0,2,Fitness,18.279451,-66.037064,Pacific Asia,Bikaner,India,19492,1/13/2018 12:27,75939,1360,16.389999,0.05,179254,327.75,-0.80,1,327.75,311.359985,-249.089996,South Asia,Rajastán,PENDING,NaN,1360,73,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,1/18/2018 12:27,Standard Class
2,CASH,4,4,-247.779999,309.720001,Shipping on time,0,73,Sporting Goods,San Jose,EE. UU.,XXXXXXXXX,Gillian,19491,Maldonado,XXXXXXXXX,Consumer,CA,8510 Round Bear Gate,95125.0,2,Fitness,37.292233,-121.881279,Pacific Asia,Bikaner,India,19491,1/13/2018 12:06,75938,1360,18.030001,0.06,179253,327.75,-0.80,1,327.75,309.720001,-247.779999,South Asia,Rajastán,CLOSED,NaN,1360,73,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,1/17/2018 12:06,Standard Class
3,DEBIT,3,4,22.860001,304.809998,Advance shipping,0,73,Sporting Goods,Los Angeles,EE. UU.,XXXXXXXXX,Tana,19490,Tate,XXXXXXXXX,Home Office,CA,3200 Amber Bend,90027.0,2,Fitness,34.125946,-118.291016,Pacific Asia,Townsville,Australia,19490,1/13/2018 11:45,75937,1360,22.940001,0.07,179252,327.75,0.08,1,327.75,304.809998,22.860001,Oceania,Queensland,COMPLETE,NaN,1360,73,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,1/16/2018 11:45,Standard Class
4,PAYMENT,2,4,134.210007,298.250000,Advance shipping,0,73,Sporting Goods,Caguas,Puerto Rico,XXXXXXXXX,Orli,19489,Hendricks,XXXXXXXXX,Corporate,PR,8671 Iron Anchor Corners,725.0,2,Fitness,18.253769,-66.037048,Pacific Asia,Townsville,Australia,19489,1/13/2018 11:24,75936,1360,29.500000,0.09,179251,327.75,0.45,1,327.75,298.250000,134.210007,Oceania,Queensland,PENDING_PAYMENT,NaN,1360,73,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,1/15/2018 11:24,Standard Class


In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 180519 entries, 0 to 180518
Data columns (total 53 columns):
 #   Column                         Non-Null Count   Dtype  
---  ------                         --------------   -----  
 0   Type                           180519 non-null  object 
 1   Days for shipping (real)       180519 non-null  int64  
 2   Days for shipment (scheduled)  180519 non-null  int64  
 3   Benefit per order              180519 non-null  float64
 4   Sales per customer             180519 non-null  float64
 5   Delivery Status                180519 non-null  object 
 6   Late_delivery_risk             180519 non-null  int64  
 7   Category Id                    180519 non-null  int64  
 8   Category Name                  180519 non-null  object 
 9   Customer City                  180519 non-null  object 
 10  Customer Country               180519 non-null  object 
 11  Customer Email                 180519 non-null  object 
 12  Customer Fname                

In [8]:
df['Order Id'].nunique()

65752

In [9]:
#drop redundant column
cols_to_drop = ['Customer Email', 'Customer Fname', 'Customer Lname', 'Customer Password', 'Customer Street', 'Product Description', 'Order Zipcode', 'Category Id', 'Product Category Id', 'Order Item Cardprod Id', 'Department Id', 'Product Image']

df = df.drop(columns = cols_to_drop)
df.shape

(180519, 41)

In [13]:
#change date type
df['order date (DateOrders)'] = pd.to_datetime(df['order date (DateOrders)'])
df['shipping date (DateOrders)'] = pd.to_datetime(df['shipping date (DateOrders)'])
                                                                                           

In [14]:
df.dtypes[['order date (DateOrders)', 'shipping date (DateOrders)']]    

order date (DateOrders)       datetime64[ns]
shipping date (DateOrders)    datetime64[ns]
dtype: object

In [15]:
#verify the shipping date
df['calculated_shipping_days'] = (df['shipping date (DateOrders)'] - df['order date (DateOrders)']).dt.days
df[['order date (DateOrders)', 'shipping date (DateOrders)', 'calculated_shipping_days', 'Days for shipping (real)']].head()

,order date (DateOrders),shipping date (DateOrders),calculated_shipping_days,Days for shipping (real)
0,2018-01-31 22:56:00,2018-02-03 22:56:00,3,3
1,2018-01-13 12:27:00,2018-01-18 12:27:00,5,5
2,2018-01-13 12:06:00,2018-01-17 12:06:00,4,4
3,2018-01-13 11:45:00,2018-01-16 11:45:00,3,3
4,2018-01-13 11:24:00,2018-01-15 11:24:00,2,2


In [20]:
#check yes or no if calculated shipping days is same as days for shipping
(df['calculated_shipping_days'] == df['Days for shipping (real)']).all()


np.False_

In [25]:
mismatches = df['calculated_shipping_days'] != df['Days for shipping (real)']
mismatches.sum()

#after two check, we see that calculated_shippingdays is not good

np.int64(4657)

In [28]:
mismatches = df['calculated_shipping_days'] != df['Days for shipping (real)']
df[mismatches][['order date (DateOrders)', 'shipping date (DateOrders)', 'calculated_shipping_days', 'Days for shipping (real)']].head(10)
                                                                                                                        

,order date (DateOrders),shipping date (DateOrders),calculated_shipping_days,Days for shipping (real)
38,2018-01-12 23:30:00,2018-01-13 11:30:00,0,1
39,2018-01-12 23:09:00,2018-01-13 11:09:00,0,1
40,2018-01-12 22:48:00,2018-01-13 10:48:00,0,1
41,2018-01-12 22:27:00,2018-01-13 10:27:00,0,1
447,2018-01-11 14:13:00,2018-01-12 02:13:00,0,1
448,2018-01-11 13:52:00,2018-01-12 01:52:00,0,1
449,2018-01-11 13:31:00,2018-01-12 01:31:00,0,1
450,2018-01-11 13:10:00,2018-01-12 01:10:00,0,1
451,2018-01-11 12:49:00,2018-01-12 00:49:00,0,1
452,2018-01-11 12:28:00,2018-01-12 00:28:00,0,1


In [29]:
#confirm the difference is by definition, not just error
diff = df.loc[mismatches, 'Days for shipping (real)'] - df.loc[mismatches, 'calculated_shipping_days']
diff.value_counts()

1    4657
Name: count, dtype: int64

Verified Days for shipping (real) against a manually calculated shipping duration. 2.6% of rows (4,657/180,519) sho a discrepancy of exactly 1 day, consistently in the same direction - caused by the original column counting calendar days crossed rather than elapsed 24-hour periods(e.g., an order place 11:30 PM and shipped 11:30 AM the next day spans 2 calendar dates but only 12 actual hours). both are valud definitions; we retain the original Days for shipping (real) column since it's what the Late_delivery_risk label was presumably built against, and drop our calculated column

In [31]:
#drop calculated_shipping_days
df = df.drop(columns = ['calculated_shipping_days'])
df.shape

(180519, 41)

In [32]:
df['Order City'].nunique()

3597

Order City retained despite high cardinality (3,597 unique values) - useful for dashboard drill-downs (top cities by order volume/ late-delivery rate/ product category), but will be excluded from delivery-risk model features due to cardinality

In [35]:
df[df['Customer Zipcode'].isna()]

,Type,Days for shipping (real),Days for shipment (scheduled),Benefit per order,Sales per customer,Delivery Status,Late_delivery_risk,Category Name,Customer City,Customer Country,Customer Id,Customer Segment,Customer State,Customer Zipcode,Department Name,Latitude,Longitude,Market,Order City,Order Country,Order Customer Id,order date (DateOrders),Order Id,Order Item Discount,Order Item Discount Rate,Order Item Id,Order Item Product Price,Order Item Profit Ratio,Order Item Quantity,Sales,Order Item Total,Order Profit Per Order,Order Region,Order State,Order Status,Product Card Id,Product Name,Product Price,Product Status,shipping date (DateOrders),Shipping Mode
35704,DEBIT,2,1,66.379997,189.660004,Late delivery,1,Consumer Electronics,CA,EE. UU.,14577,Consumer,95758,NaN,Technology,39.134159,-123.718552,Europe,Valence,Francia,14577,2017-11-02 18:31:00,71024,63.220001,0.25,174339,252.880005,0.35,1,252.880005,189.660004,66.379997,Western Europe,Auvernia-Ródano-Alpes,COMPLETE,1352,Industrial consumer electronics,252.880005,0,2017-11-04 18:31:00,First Class
46440,PAYMENT,4,4,10.910000,38.959999,Shipping on time,0,Video Games,CA,EE. UU.,17171,Corporate,95758,NaN,Discs Shop,38.408573,-121.449112,Pacific Asia,Manila,Filipinas,17171,2017-12-10 15:18:00,73618,0.800000,0.02,176933,39.750000,0.28,1,39.750000,38.959999,10.910000,Southeast Asia,Capital Nacional,PENDING_PAYMENT,1362,Fighting video games,39.750000,0,2017-12-14 15:18:00,Standard Class
82511,DEBIT,4,4,59.990002,299.959992,Shipping on time,0,Children's Clothing,CA,EE. UU.,14046,Consumer,91732,NaN,Apparel,34.071892,-118.013886,Europe,Sheffield,Reino Unido,14046,2017-10-26 00:29:00,70493,57.139999,0.16,173808,357.100006,0.20,1,357.100006,299.959992,59.990002,Northern Europe,Inglaterra,COMPLETE,1350,Children's heaters,357.100006,0,2017-10-30 00:29:00,Standard Class


In [37]:
#drop the row with NaN customer zipcode
df = df.dropna(subset = ['Customer Zipcode'])
df.shape

(180516, 41)

Customer Zipcode has three entries that are NaN. Those 3 entries might have been a misinterpretation for other the LATAM market (as they write EE.UU instead of United States). The entries made up of only 0.0017% of the data, instead of trying to guess correctly and reconstruct the zipcode, we decided to get rid of the entries
# Better version
Customer Zipcode is missing for 3 rows (0.0017% of the data). Inspecting these rows also revealed apparent column misalignment in the geography fields (e.g., Customer State holding what looks like a zip code, Customer Country showing "EE.UU." — Spanish for "United States," suggesting a LATAM-market entry issue). Given the negligible sample size and that correcting this would require guessing rather than recovering real values, these 3 rows were dropped rather than imputed.

### Clearning Summary
Dropped 12 unusable/PII/redundant columns (53 -> 14)
Converted both date and columns to real datetimes
verified Days for shipping (real) against a manual calculation, found and explained a calendar-day-vs-24 hour decrepancy, kept the original column
decided to keep order city despite high cardinality, with a note to exclude it from model features later
dropped 3 rows with missing zipcode/misaligned geography data, with a documented reason

In [ ]:
df.to_csv('../data/processed/cleaned_supply_chain.csv', index = False)